apis/goharbor.io/v1alpha2/chartmuseum_types.go

```diff
@@ -0,0 +1,362 @@
+	Provenance string `json:"provenance,omitempty"`
+}
+
+type ChartMuseumLogSpec struct {
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=false
+	// Output structured logs as json
+	JSON bool `json:"json"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=false
+	// Show debug messages
+	Debug bool `json:"debug"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=false
+	// Log inbound /health requests
+	Health bool `json:"health"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=true
+	// log latency as an integer instead of a string
+	LatencyInteger *bool `json:"latencyInteger,omitempty"`
+}
+
+type ChartMuseumAuthSpec struct {
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=false
+	// Allow anonymous GET operations when auth is used
+	AnonymousGet bool `json:"anonymousGet"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Pattern="[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*"
+	// Reference to secret containing basic http authentication
+	BasicAuthRef string `json:"basicAuthRef,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// Bearer authentication specs
+	Bearer *ChartMuseumAuthBearerSpec `json:"bearer,omitempty"`
+}
+
+type ChartMuseumAuthBearerSpec struct {
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern="[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*"
+	// Reference to secret containing authorization server certificate
+	CertificateRef string `json:"certificateRef"`
+
+	// +kubebuilder:validation:Required
+	// Authorization server url
+	Realm string `json:"realm"`
+
+	// +kubebuilder:validation:Required
+	// Authorization server service name
+	Service string `json:"service"`
+}
+
+type ChartMuseumHTTPSSpec struct {
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern="[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*"
+	// Reference to secret containing tls certificate
+	CertificateRef string `json:"certificateRef"`
+}
+
+type ChartMuseumDisableSpec struct {
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=false
+	// Disable all routes prefixed with
+	API bool `json:"api"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=false
+	// Disable use of index-cache.yaml
+	StateFiles bool `json:"statefiles"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=false
+	// Do not allow chart versions to be re-uploaded, even with ?force querystrin
+	ForceOverwrite bool `json:"forceOverwrite"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=false
+	// Disable Prometheus metrics
+	Metrics bool `json:"metrics"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=false
+	// Disable DELETE route
+	Delete bool `json:"delete"`
+}
+
+type ChartMuseumCacheSpec struct {
+	// +kubebuilder:validation:Optional
+	// Redis cache store
+	Redis *OpacifiedDSN `json:"redis,omitempty"`
```

Comment:
Why does each of the components still need to specify their own Redis configuration? Is it possible to let the user specified a global Redis in the root spec and populate the Redis conn to each of the required components with different database indexes?

Reply:
Each components have the redis spec. But controlled by the Harbor resource which have a common redisDSN.
So defining redis through the Harbor spec will populate all other resource with the right redis database.
See in the make file. We now have a single redis deployment.
And I am currently working on the harbor sample to make it works.

Reply:
understand now